Do not select all entities when omitting entity ID in service call

balloob opened the pull request:
## Breaking Change:

It is no longer possible to target all entities by not passing both `entity_id` and `area_id` to your service data. This feature was previously deprecated and has been printing warnings. The correct way to target all entities is to set `entity_id: all`

## Description:
Remove the feature to automatically target all entities if you did not specify an entity. 

Previously attempted in #25715.

`entity_id` has not been set to required because it is allowed to just specify `area_id` to select entities.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

balloob committed:
Do not select all entities when omitting entity ID

amelchio commented on homeassistant/helpers/service.py:
```diff
@@ -260,19 +260,7 @@ def async_set_service_schema(hass, domain, service, schema):
-    if ATTR_ENTITY_ID in call.data:
-        target_all_entities = call.data[ATTR_ENTITY_ID] == ENTITY_MATCH_ALL
-    else:
-        # Remove the service_name parameter along with this warning
```
This comments says that further changes are needed if it is deleted :)

amelchio reviewed (commented):
> `entity_id` has not been set to required because it is allowed to just specify `area_id` to select entities.

So we now fail silently?

It would be quite useful to have the constraint (need one or the other) encoded in the schemas.

pvizeli commented on homeassistant/helpers/service.py:
```diff
@@ -260,19 +260,7 @@ def async_set_service_schema(hass, domain, service, schema):
-            service_name,
-            ENTITY_MATCH_ALL,
-        )
-        target_all_entities = True
```
We should remove the auto target and convert the warning into error and fail on this service call

MatthewFlamm commented on tests/helpers/test_entity_component.py:
```diff
@@ -259,7 +260,7 @@ def platform_setup(hass, config, add_entities, discovery_info=None):
     call = ha.ServiceCall("test", "service")
 
-    assert ["test_domain.test_1", "test_domain.test_2"] == sorted(
+    assert [] == sorted(
```
The name of this test probably needs to be changed now.

balloob replied:
Fixed

MatthewFlamm commented on tests/helpers/test_service.py:
```diff
@@ -470,13 +475,7 @@ def test_async_get_all_descriptions(hass):
-        "Not passing an entity ID to a service to target " "all entities is deprecated"
-    ) in caplog.text
+    assert entities == []
 
```
Based on title of test, should this one use `ENTITY_MATCH_ALL`

MatthewFlamm replied:
Sorry for leaving slightly misplaced comments, I couldn't figure out how to leave comments in parts of unchanged code.

MatthewFlamm replied:
Still can't figure it out, in the previous test, you can use `ENTITY_MATCH_ALL` instead of `"all"`

balloob replied:
ENTITY_MATCH_ALL is defined as `all` :)

MatthewFlamm replied:
My comment was around consistency, but you are right that it doesn't matter for functionality.  Just a suggestion.

balloob committed:
Address comments Matthew

balloob commented:
I've updated the code to always require either `entity_id` or `area_id`. I did some clean up in the process. The change was made by making sure that ENTITY_SERVICE_SCHEMA has been removed and replaced with a function that will extend the passed in schema, but also enforced either key is passed in.

balloob committed:
Require either area_id or entity_id

balloob committed:
Fix tests

balloob committed:
Fix test

eman291 commented:
Duplicate of #
Duplicate of #27865 ](url)

eman291 commented:
@smurfix 
@ #29292
